### ISSUE TITLE
reconsider constraints on operator overloading

### ISSUE THREAD

**Issue by maddanio** [vlang/v#11304]
Hello,
ally for scientific applications, but then you constrain it to be type-symmetric, i.e. dont allow operators between different types. Unfortunately that seems to strongly constrain scientific application to me. Matrices, vectors, scalars, tensors, quaternions, cokmplex numbers, all of these define cross-type operators that are very useful and otherwise lead to a lot of what the doc says it doesnt want (i.e. matrix.dot(vector).mul(scalar)) as can be seen in vlibs implementation of glm.

**Comment by medvednikov** [comment]
I agree, that would clean up the code significantly. But in V like in Go all operators must have the same types on both sides.

Function overloading is also not allowed, so having `vec * vec` and `vec * int` would violate that rule as well.

**Comment by maddanio** [comment]
Operators are alteady an exception to this rule, my suggestion is to widen that exception so the claimed goal, supporting scientific computing, can be reached

**Comment by medvednikov** [comment]
How are they an exception? They types must be the same: `int + int`, `string + string` etc.

**Comment by medvednikov** [comment]
Oh you mean operator overloading being like function overloading.

Well not really, the name is similar, but operator overloading is basically defining new functions with syntax sugar, not overloading them.

**Comment by maddanio** [comment]
not really, + is basically a function name like any other, and you are overloading it :). anyway, I just thought since in the motivation to allow it as an exception (as which it is also documented) as stated in the documentation was to support scientific computing, and not allowing asymmetry really doesnt do much in that direction, since all the interesting operators are asymmetric, but feel free to close this.

btw, is there any kind of decision process on such things? who gets the final say? just wondering. as far as I see you are the one doing the heavy lifting?

**Comment by medvednikov** [comment]
No, you are not overloading, you are creating a new function.

Vec + Vec is not defined in V, you are defining a new method `plus`.

**Comment by maddanio** [comment]
Ah, as a member, ok, it's usually called operator overloading, and in my mind i usually make little distinction between methods and functions. And the vlang doc also calls it overloading :). And the text basically says its an exception toward the simplicity philosophy for scientific applications. I think in this limited way it doesn't achieve that, but thats just me.
Still curious if there is any kind of design decision process:)

**Comment by medvednikov** [comment]
Like I said, it's called operator, overloading, but nothing is being overloaded, like with function overloading. In V methods are functions with receivers, syntax sugar.

**Comment by maddanio** [comment]
Sure :). Didnt really want to argue semantics though
